app/Models/TagModel.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace App\Models;
+
+use App\Entities\Tag;
+use CodeIgniter\Model;
+use ReflectionException;
+
+class TagModel extends Model
+{
+    protected $table          = 'tags';
+    protected $primaryKey     = 'id';
+    protected $returnType     = Tag::class;
+    protected $useSoftDeletes = false;
+    protected $protectFields  = true;
+    protected $allowedFields  = ['name'];
+    protected $useTimestamps  = true;
+
+    /**
+     * Create many tags
+     */
+    public function createTags(array $tags, int $threadId): void
+    {
+        foreach ($tags as $tag) {
+            $this->createTag($tag, $threadId);
+        }
+    }
+
+    /**
+     * Update many tags
+     */
+    public function updateTags(array $tags, int $threadId): void
+    {
+        $this->db->table('thread_tags')->where('thread_id', $threadId)->delete();
+
+        $this->createTags($tags, $threadId);
+    }
+
+    /**
+     * Create tag
+     */
+    public function createTag(string $name, int $threadId): int
+    {
+        $data = [
+            'tag_id'    => $this->findOrCreateId($name),
+            'thread_id' => $threadId,
+        ];
+
+        $this->db->table('thread_tags')->insert($data);
+
+        return $data['tag_id'];
+    }
+
+    /**
+     * Find or create tag and return tag ID.
+     *
+     * @throws ReflectionException
+     */
+    public function findOrCreateId(string $name): int
+    {
+        $tagId = $this->where('name', $name)->first()?->id;
```

Comment:
Should we standardize all tags to lowercase? thinking maybe we should since the name is used in the URI.

Reply:
Unless we don't want to be so restrictive and allow uppercase and then convert everything to lowercase...

Reply:
Ok, I missed that! I think that's fine for now. We can always update later if it gets to feeling restrictive.